core/src/main/java/com/linecorp/armeria/client/routing/ZookeeperEndpointGroup.java

```
@@ -0,0 +1,154 @@
+ */
+public class ZookeeperEndpointGroup implements EndpointGroup, Watcher, StatCallback {
+
+    private static final Logger logger = LoggerFactory.getLogger(ZookeeperEndpointGroup.class);
+
+    private String hostPort;
+    private String znode;
+    private List<Endpoint> endpointList;
+    private ZooKeeper zooKeeper;
+    byte[] prevData;
+    Function<byte[], List<Endpoint>> nodeValueToEndpointList;
+
+    final CountDownLatch connectionLatch = new CountDownLatch(1);
+
+    /**
+     * create a zookeeper endpoint group
+     * @param hostPort zookeeper hostPort eg. localhost:2181
+     * @param zNode a zNode path eg. /groups/productionGroups
+     * @param sessionTimeout   zookeeper session timeout
+     * @param nodeValueToEndpointList a function to convert zNode value to a List of {@link Endpoint}
+     */
+    public ZookeeperEndpointGroup(String hostPort, String zNode, int sessionTimeout,
+                                  Function<byte[], List<Endpoint>> nodeValueToEndpointList) {
+
+        requireNonNull(hostPort);
+        requireNonNull(zNode);
+        requireNonNull(nodeValueToEndpointList);
+        this.hostPort = hostPort;
+        this.znode = zNode;
+        this.nodeValueToEndpointList = nodeValueToEndpointList;
+        try {
+            zooKeeper = new ZooKeeper(hostPort, sessionTimeout, event -> {
+                if (event.getState() == KeeperState.SyncConnected) {
+                    connectionLatch.countDown();
+                }
+
+            });
+            connectionLatch.await();
+
+            Stat stat = zooKeeper.exists(zNode, this);
+            byte[] nodeData;
+            if (stat != null) {
+                nodeData = zooKeeper.getData(zNode, false, null);
+                endpointList = nodeValueToEndpointList.apply(nodeData);
+            }
+
+        } catch (IOException e) {
+            logger.error("error occurred while connecting zooKeeper " + e.getMessage());
+            e.printStackTrace();
```

Instead of logging, please throw a new `EndpointGroupException`. For example:

```java
throw new EndpointGroupException("failed to connect to ZooKeeper: " + hostPort, e);
```